### PR TITLE
Build: multi arch images

### PR DESCRIPTION
**Pull request by morlay**
**What this PR does**:

Build: multi arch images (amd64 && arm64)

**Which issue(s) this PR fixes**:
No fixes

**Checklist**
- [x] <del>waiting for upsteam `jaegertracing/jaeger-query` </del> custom build for `jaegertracing/jaeger-query`
- [x] CI updates, may switch to use github actions

Local run (need to install buildx https://github.com/docker/buildx#installing)

```bash
NAMESPACE=$(whoami) make publish-docker-images
```
examples:

https://hub.docker.com/r/morlay/tempo/tags
https://hub.docker.com/r/morlay/tempo-vulture/tags
https://hub.docker.com/r/morlay/tempo-query/tags (a custom build stage for jaegertracing/jaeger-query)

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -13,6 +13,12 @@ jobs:
         go-version: 1.15
       id: go
 
+    - name: Set up qemu
+      uses: docker/setup-qemu-action@v1
+
+    - name: Set up buildx
+      uses: docker/setup-buildx-action@v1
+
```
Comment (morlay) [inline .github/workflows/ci.yml:21]: setup buildx for multi-arch docker build

https://github.com/docker/setup-buildx-action

File: Makefile [hunk 2]
```diff
@@ -4,6 +4,13 @@ VERSION=$(shell ./tools/image-tag | cut -d, -f 1)
 GIT_REVISION := $(shell git rev-parse --short HEAD)
 GIT_BRANCH := $(shell git rev-parse --abbrev-ref HEAD)
 
+NAMESPACE ?= grafana
```
Comment (morlay) [inline Makefile:7]: for docker publish target

File: Makefile [hunk 3]
```diff
@@ -4,6 +4,13 @@ VERSION=$(shell ./tools/image-tag | cut -d, -f 1)
...
+BUILDX_OUTPUT ?= --load
+PLATFORMS ?= linux/amd64
+
+GOOS ?= $(shell go env GOOS)
+GOARCH ?= $(shell go env GOARCH)
+
```
Comment (morlay) [inline Makefile:13]: `GOARCH` for switching build arch target

File: Makefile [hunk 4]
```diff
@@ -35,24 +42,24 @@ endif
 
 .PHONY: tempo
 tempo:
-	GO111MODULE=on CGO_ENABLED=0 go build $(GO_OPT) -o ./bin/$(GOOS)/tempo $(BUILD_INFO) ./cmd/tempo
+	GO111MODULE=on CGO_ENABLED=0 go build $(GO_OPT) -o ./bin/$(GOOS)/tempo-$(GOARCH) $(BUILD_INFO) ./cmd/tempo
```
Comment (morlay) [inline Makefile:45]: arch suffix

File: Makefile [hunk 5]
```diff
@@ -35,24 +42,24 @@ endif
...
 tempo-vulture:
-	GO111MODULE=on CGO_ENABLED=0 go build $(GO_OPT) -o ./bin/$(GOOS)/tempo-vulture $(BUILD_INFO) ./cmd/tempo-vulture
+	GO111MODULE=on CGO_ENABLED=0 go build $(GO_OPT) -o ./bin/$(GOOS)/tempo-vulture-$(GOARCH) $(BUILD_INFO) ./cmd/tempo-vulture
 
 .PHONY: exe
 exe:
-	GOOS=linux $(MAKE) $(COMPONENT)
-	cp ./bin/linux/$(COMPONENT) ./cmd/$(COMPONENT)/
+	GOOS=$(GOOS) $(MAKE) $(COMPONENT)
+	cp ./bin/$(GOOS)/$(COMPONENT)-$(GOARCH) ./cmd/$(COMPONENT)/$(COMPONENT)
```
Comment (morlay) [inline Makefile:62]: just make exe no breaking changes.

File: Makefile [hunk 6]
```diff
@@ -80,10 +87,8 @@ lint:
 ### Docker Images
 
 .PHONY: docker-component # Not intended to be used directly
-docker-component: check-component exe
-	docker build -t grafana/$(COMPONENT) -f ./cmd/$(COMPONENT)/Dockerfile .
-	docker tag grafana/$(COMPONENT) $(COMPONENT)
-	rm ./cmd/$(COMPONENT)/$(COMPONENT)
+docker-component: check-component
```
Comment (morlay) [inline Makefile:90]: build go binary in Dockerfile. 
so no need exe to build again.

File: Makefile [hunk 7]
```diff
@@ -80,10 +87,8 @@ lint:
...
+	docker buildx build $(BUILDX_OUTPUT) --platform=$(PLATFORMS) -t $(NAMESPACE)/$(COMPONENT) -f ./cmd/$(COMPONENT)/Dockerfile .
```
Comment (morlay) [inline Makefile:91]: docker tag alias here have issue for when set `BUILDX_OUTPUT=--push`
means no local image for tag.

File: integration/e2e/util.go [hunk 8]
```diff
@@ -23,7 +23,7 @@ import (
 )
 
 const (
-	image = "tempo:latest"
+	image = "grafana/tempo:latest"
```
Comment (morlay) [inline integration/e2e/util.go:26]: no tag alias any more. so need using fullname

File: cmd/tempo-query/Dockerfile [hunk 9]
```diff
@@ -1,8 +1,62 @@
...
+
+FROM --platform=${BUILDPLATFORM:-linux/amd64} golang:1.15.3-alpine AS builder
+
+RUN apk --update add make git
+
+WORKDIR /go/src
+COPY ./ ./
+
+ARG TARGETARCH
+RUN GOARCH=${TARGETARCH} make tempo-query
```
Comment (morlay) [inline cmd/tempo-query/Dockerfile:52]: when `--platform=${BUILDPLATFORM:-linux/amd64}` in `FROM`

mean as (only on amd64 env)
```bash
for TARGETARCH in "amd64 arm64"; do
    GOARCH=${TARGETARCH} make tempo-query
done
```

**Comment by morlay** [comment]
@joe-elliott @annanay25 
could you mind switching release job to github actions?
i have no idea to setup it on drone

File: .github/workflows/publish-images.yml [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+          GITHUB_SHA=${{ github.sha }}
+
+          IMAGES=${HUB}/${{ matrix.component }}:latest,${HUB}/${{ matrix.component }}:${GITHUB_SHA::8}
+
+          if [[ ${{ github.ref }} == "refs/tags/v"* ]]; then
+            GITHUB_REF=${{ github.ref }}
+            IMAGES=${HUB}/${{ matrix.component }}:${GITHUB_REF/refs\/tags\/v/}
+          fi
+
+          echo ::set-output name=image::${IMAGES//$','/'%0A'}
```
Comment (morlay) [inline .github/workflows/publish-images.yml:39]: to pass multi line image tag list with `\n`

https://github.community/t/set-output-truncates-multiline-strings/16852

**Comment by joe-elliott** [comment]
So I generally agree that building arm64 images is desirable, but I have some concerns about this PR:

- I don't like that the default `make docker-*` commands use buildx as it is an experimental docker feature and requires turning on a feature flag.
- We shouldn't be building Jaeger-query from scratch.  We should wait for Jaeger to build arm64 images and use those.  I saw that you were working on a PR there and I'm fine with waiting for that to be complete.
- Currently we build the binary outside the Dockerfile and just copy it in.  This is very fast and I'd prefer continuing to do this.

Also, unfortunately we can't switch off of drone as Grafana secrets are only available in our private drone cluster.  I too wish we could stick with just Github Actions.

Also also, this PR is a lot to review and I think we should break it up.  Just a `make` command to build other architectures and the necessary supporting files would be a good place to start.  We can worry about CI later.

**Comment by morlay** [comment]
@joe-elliott 

Sure. i will open another pr just upgrade Makefile to make multi arch build happy.

**Pull request closed without merging**